### DISCUSSION TITLE
Похожий плейлист

### DISCUSSION THREAD

**Discussion by Bedmit** [Chimildic/goofy#88] [Помощь с алгоритмом] [answered]
Прошу помощи в редактировании скрипта. И так на половину не мной составленного ))
Скрипт генерирует похожий плейлист функцией `replaceWithSimilar` по ссылке на плейлист, вставленной в мобильную форму. Если ссылка не введена, то обрабатывается последняя вставленная до этого.
```
function Fresh(values) { 
  const LastFrech = 'LastPlaylistFreshFunc.txt'
  let playlistId;
  if (values.text.playlistUrl.length == 0) {
    playlistId = Cache.read(LastFrech);
  } else {
    playlistId = Helper.parseId(values.text.playlistUrl);  
    Cache.write(LastFrech, playlistId);
  }  

  let Fresh = Source.getPlaylistTracks('', playlistId);
  Filter.replaceWithSimilar(Fresh, Source.getSavedTracks(), RecentTracks.get());
  Filter.matchExcept(Fresh, 'mix|club|radio|edit|live|karaoke');
  Filter.dedupTracks(Fresh);

  let playlist = Playlist.getById(playlistId);
  
  Playlist.saveAsNew({
    name: playlist.name + ' *',
    description: 'По мотивам ' + playlist.name,
    public: false,
    sourceCover: playlist.images[0].url,
    tracks: Fresh,  
  });  
}
```
Нужно сделать так, чтобы треки которые остались от плейлиста-источника и не изменились, шли в начале полученного плейлиста, в той же очередности. А после них уже шли треки заменённые на похожие.

**Comment by Chimildic** [comment, marked as answer]
```js
  // ...

  let Fresh = Source.getPlaylistTracks('', playlistId);
  let FreshCopy = Selector.sliceCopy(Fresh);
  Filter.replaceWithSimilar(Fresh, Source.getSavedTracks(), RecentTracks.get());
  Filter.matchExcept(Fresh, 'mix|club|radio|edit|live|karaoke');
  Filter.dedupTracks(Fresh);

  Filter.removeTracks(FreshCopy, Fresh, true);
  Filter.removeTracks(Fresh, FreshCopy);
  let tracks = [];
  Combiner.push(tracks, FreshCopy, Fresh);

  let playlist = Playlist.getById(playlistId);

  Playlist.saveAsNew({
    name: playlist.name + ' *',
    description: 'По мотивам ' + playlist.name,
    public: false,
    sourceCover: playlist.images[0].url,
    tracks: tracks,  
  }); 
```

**Reply by Bedmit** [reply]
Спасибо. Думал примерно в том же ключе. Только один момент не понимаю - что делает первая строка в
  ```
let tracks = [];
  Combiner.push(tracks, FreshCopy, Fresh);
```
Я думал вообще можно обойтись только `Combiner.push(FreshCopy, Fresh);`

**Reply by Chimildic** [reply]
Да, для текущей задачи выглядит лишним. Просто для читаемости, отслеживания логики, создается новая переменная. 
Знак `[]` - означает создать пустой массив.

**Reply by Bedmit** [reply]
Спасибо, понял. Понимать бы еще в каких случаях это применять, а в каких нет. Надеюсь со временем дойдет.

**Reply by Chimildic** [reply]
Очень многие решение касающиеся реализации, а не логики могут запутать. Такие решения приходят из опыта, из желания повысить читаемость или просто табу на стиль написания. Не стоит на них зацикливаться. Одну и ту же логику можно написать по-разному.
